File: lib/gui/tray_icon.py
import os
import signal
import subprocess

# pylint: disable=E0611, E0401
from PyQt5.QtWidgets import QSystemTrayIcon, QMenu, QAction, QActionGroup, QColorDialog
from PyQt5.QtGui import QIcon, QColor

from lib.gui.game_list import GameList
from lib.gui.main_window import main_window

from lib import __app_name__, __version__
from lib.models.boost import Boost
from lib.models.thermal_throttle_profile import ThermalThrottleProfile
from lib.models.rgb_brightness import RgbBrightness
from lib.models.battery_threshold import BatteryThreshold
from lib.services.games_service import games_service
from lib.services.openrgb_service import open_rgb_service
from lib.services.platform_service import platform_service
from lib.utils.constants import icons_path, log_file, dev_mode
from lib.utils.event_bus import event_bus
from lib.utils.logger import Logger
from lib.utils.singleton import singleton
from lib.utils.translator import translator


@singleton
class TrayIcon:  # pylint: disable=R0902
    """Tray icon class"""

    def __init__(self):  # pylint: disable=R0915
        self._logger = Logger()

        icon = QIcon.fromTheme(os.path.join(icons_path, "icon-45x45.png"))

        self._tray = QSystemTrayIcon()
        self._tray.setIcon(icon)
        self._tray.setToolTip(f"{__app_name__} v{__version__}")

        # Create the menu
        self._menu = QMenu()

        # Add "Battery" option (disabled)
        self._battery_action = QAction(translator.translate("battery"))
        self._battery_action.setEnabled(False)
        self._menu.addAction(self._battery_action)

        # Add "Umbral" submenu
        self._umbral_menu = QMenu("    " + translator.translate("charge.threshold"))
        self._threshold_group = QActionGroup(self._umbral_menu)
        self._threshold_group.setExclusive(True)
        self.threshold_actions: dict[BatteryThreshold, QAction] = {}
        for threshold in BatteryThreshold:
            action = QAction(f"{threshold.value}%", checkable=True)
            action.setActionGroup(self._threshold_group)
            action.setChecked(threshold == platform_service._battery_charge_limit)
            action.triggered.connect(lambda _, t=threshold: self.on_threshold_selected(t))
            self.threshold_actions[threshold] = action
            self._umbral_menu.addAction(action)
        self._menu.addMenu(self._umbral_menu)

        self._menu.addSeparator()

        # Add "AuraSync" option (disabled)
        self._aura_section = QAction("AuraSync")
        self._aura_section.setEnabled(False)
        self._menu.addAction(self._aura_section)

        # Add "Effect" submenu
        self._effect_menu = QMenu("    " + translator.translate("effect"))
        self._effect_group = QActionGroup(self._effect_menu)
        self._effect_group.setExclusive(True)
        self._effect_actions: dict[str, QAction] = {}
        for effect in open_rgb_service.get_available_effects():
            action = QAction(effect, checkable=True)
            action.setActionGroup(self._effect_group)
            action.setChecked(effect == open_rgb_service._effect)
            action.triggered.connect(lambda _, e=effect: self.on_effect_selected(e))
            self._effect_actions[effect] = action
            self._effect_menu.addAction(action)
        self._menu.addMenu(self._effect_menu)

        # Add "Brightness" submenu
        self._brightness_menu = QMenu("    " + translator.translate("brightness"))
        self._brightness_group = QActionGroup(self._brightness_menu)
        self._brightness_group.setExclusive(True)
        self._brightness_actions: dict[str, RgbBrightness] = {}
        for brightness in RgbBrightness:
            action = QAction(
                translator.translate(f"label.brightness.{brightness.name}"),
                checkable=True,
            )
            action.setActionGroup(self._brightness_group)
            action.setChecked(brightness == open_rgb_service._brightness)
            action.triggered.connect(lambda _, b=brightness: self.on_brightness_selected(b))
            self._brightness_actions[brightness] = action
            self._brightness_menu.addAction(action)
        self._menu.addMenu(self._brightness_menu)

        # Add "Color" submenu
        self._color_menu = QMenu("    " + translator.translate("color"))

        self._color_action = QAction(open_rgb_service.get_color())
        self._color_action.setEnabled(False)
        self._color_menu.addAction(self._color_action)

        self._colorpicker_action = QAction(f"{translator.translate("select.color")}...")
        self._colorpicker_action.triggered.connect(self.pick_color)
        self._color_menu.addAction(self._colorpicker_action)

        self._submenu_color = self._menu.addMenu(self._color_menu)
        self._submenu_color.setVisible(open_rgb_service.supports_color())

        self._menu.addSeparator()

        # Add "Performance" option
        self._performance_section = QAction(translator.translate("performance"))
        self._performance_section.setEnabled(False)
        self._menu.addAction(self._performance_section)

        self._profile_menu = QMenu("    " + translator.translate("profile"))
        self._boost_group = QActionGroup(self._profile_menu)
        self._boost_group.setExclusive(True)
        self._performance_actions: dict[ThermalThrottleProfile, QAction] = {}
        for profile in ThermalThrottleProfile:
            action = QAction(translator.translate(f"label.profile.{profile.name}"), checkable=True)
            action.setActionGroup(self._boost_group)
            action.setChecked(profile == platform_service._thermal_throttle_profile)
            action.triggered.connect(lambda _, p=profile: self.on_profile_selected(p))
            self._performance_actions[profile] = action
            self._profile_menu.addAction(action)
        self._menu.addMenu(self._profile_menu)

        # Add "Boost" option
        self._boost_menu = QMenu("    " + translator.translate("boost"))
        self._boost_group = QActionGroup(self._boost_menu)
        self._boost_group.setExclusive(True)
        self._boost_actions: dict[Boost, QAction] = {}
        for mode in Boost:
            action = QAction(translator.translate(f"label.boost.{mode.name}"), checkable=True)
            action.setActionGroup(self._boost_group)
            action.setChecked(mode == platform_service.get_boost_mode())
            action.triggered.connect(lambda _, m=mode: self.on_boost_selected(m))
            self._boost_actions[mode] = action
            self._boost_menu.addAction(action)
        self._menu.addMenu(self._boost_menu)

        if games_service.rccdc_enabled:
            # Add "Games" option
            self._games_menu = QMenu("    " + translator.translate("games"))
            self._select_profile_action = QAction(f"{translator.translate("label.game.configure")}...")
            self._select_profile_action.triggered.connect(self.on_open_game_list)
            self._games_menu.addAction(self._select_profile_action)
            self._menu.addMenu(self._games_menu)

        self._menu.addSeparator()

        if dev_mode:
            # Add "Open log" option
            self._open_logs_action = QAction(translator.translate("open.logs"))
            self._open_logs_action.triggered.connect(self.on_open_logs)
            self._menu.addAction(self._open_logs_action)

            self._menu.addSeparator()

        # Add "Open" option
        self._open_action = QAction(translator.translate("open.ui"))
        self._open_action.triggered.connect(self.on_open)
        self._menu.addAction(self._open_action)

        self._menu.addSeparator()

        # Add "Quit" option
        self._quit_action = QAction(translator.translate("close"))
        self._quit_action.triggered.connect(self.on_quit)
        self._menu.addAction(self._quit_action)

        # Set the menu on the tray icon
        self._tray.setContextMenu(self._menu)

        event_bus.on("PlatformService.battery_threshold", self.set_battery_charge_limit)
        event_bus.on("PlatformService.boost", self.set_boost_mode)
        event_bus.on("PlatformService.thermal_throttle_profile", self.set_thermal_throttle_policy)
        event_bus.on("OpenRgbService.aura_changed", self.set_aura_state)
        event_bus.on("GamesService.gameEvent", self.on_game_event)

        self._tray.activated.connect(self.on_tray_icon_activated)

    def on_game_event(self, running_games: int):
        """Handler for game events"""
        enable = running_games == 0
        self._profile_menu.setEnabled(enable)
        self._boost_menu.setEnabled(enable)
        self._games_menu.setEnabled(enable)

    def on_tray_icon_activated(self, reason):
        """Restore main window"""
        if reason == QSystemTrayIcon.Trigger:
            self.on_open()

    def set_battery_charge_limit(self, value: BatteryThreshold):
        """Set battery charge limit"""
        self.threshold_actions[value].setChecked(True)

    def set_boost_mode(self, value: Boost):
        """Set boost mode"""
        self._boost_actions[value].setChecked(True)

    def set_thermal_throttle_policy(self, value: ThermalThrottleProfile):
        """Set performance profile"""
        self._logger.debug("Refreshing throttle policy in UI")
        self._performance_actions[value].setChecked(True)

    def set_aura_state(self, value):
        """Set aura state"""
        effect, brightness, color = value

        self._effect_actions[effect].setChecked(True)
        self._brightness_actions[brightness].setChecked(True)
        if color is None:
            self._submenu_color.setVisible(False)
        else:
            self._submenu_color.setVisible(True)
            self._color_action.setText(color)

    def show(self):
        """Show tray"""
        self._tray.setVisible(True)

    def on_threshold_selected(self, threshold: BatteryThreshold):
        """Battery limit event handler"""
        platform_service.set_battery_threshold(threshold)

    def on_effect_selected(self, effect: str):
        """Effect event handler"""
        open_rgb_service.apply_effect(effect)

    def on_brightness_selected(self, brightness: RgbBrightness):
        """Brightness event handler"""
        open_rgb_service.apply_brightness(brightness)

    def on_profile_selected(self, profile: ThermalThrottleProfile):
        """Profile event handler"""
        platform_service.set_thermal_throttle_policy(profile)

    def on_boost_selected(self, boost: Boost):
        """Boost event handler"""
        platform_service.set_boost_mode(boost)

    def pick_color(self):
        """Open color picker"""
        color = QColorDialog.getColor(QColor(self._color_action.text()), None, translator.translate("color.select"))
        if color.isValid():
            open_rgb_service.apply_color(color.name().upper())

    def on_open_game_list(self):
        """Open game list window"""
        GameList().show()

    @staticmethod
    def on_open():
        """Open main window"""
        main_window.show()

    @staticmethod
    def on_open_logs():
        """Open log file"""
        subprocess.run(["xdg-open", log_file], check=False)

    @staticmethod
    def on_quit():
        """Quit application"""
        event_bus.emit("stop", None)
        os.kill(os.getpid(), signal.SIGKILL)


tray_icon = TrayIcon()

File: RogPerfTuner/src/gui/tray_icon.cpp
#include "gui/tray_icon.hpp"

#include <qaction.h>
#include <qcolordialog.h>
#include <qkeysequence.h>
#include <qobject.h>

#include <QActionGroup>
#include <QMenu>
#include <array>
#include <optional>

#include "gui/fan_curve_editor.hpp"
#include "gui/game_list.hpp"
#include "models/performance/performance_profile.hpp"
#include "utils/enum_utils.hpp"
#include "utils/string_utils.hpp"

void TrayIcon::openMainWindow() {
	mainWindow.show();
	mainWindow.raise();
	mainWindow.activateWindow();
}

void TrayIcon::openSettings() {
	shell.run_command("xdg-open " + Constants::CONFIG_FILE);
}

void TrayIcon::reloadSettings() {
	configuration.loadConfig();
}

void TrayIcon::openLogs() {
	shell.run_command("xdg-open " + Constants::LOG_DIR + "/" + Constants::LOG_FILE_NAME + ".log");
}

void TrayIcon::openGameList() {
	if (GameList::INSTANCE == nullptr) {
		GameList::INSTANCE = new GameList(&mainWindow);
	}
	GameList::INSTANCE->show();
}

void TrayIcon::openFanEditor() {
	auto profile		= toString(performanceService.getPerformanceProfile());
	CurveEditor* editor = new CurveEditor(profile, &mainWindow);
	editor->show();
}

void TrayIcon::setAuraBrightness(RgbBrightness brightness) {
	QMetaObject::invokeMethod(
		this,
		[=, this]() {
			brightnessActions[toName(brightness)]->setChecked(true);
		},
		Qt::QueuedConnection);
}

void TrayIcon::setAuraEffect(const std::string& effect) {
	QMetaObject::invokeMethod(
		this,
		[=, this]() {
			effectActions[effect]->setChecked(true);
		},
		Qt::QueuedConnection);
}

void TrayIcon::setAuraColor(const std::optional<std::string>& color) {
	QMetaObject::invokeMethod(
		this,
		[=, this]() {
			if (color.has_value()) {
				colorMenu->setEnabled(true);
				currentColorAction->setText(StringUtils::toUpperCase(color.value()).c_str());
			} else {
				colorMenu->setEnabled(false);
			}
		},
		Qt::QueuedConnection);
}

void TrayIcon::setPerformanceProfile(PerformanceProfile profile) {
	QMetaObject::invokeMethod(
		this,
		[=, this]() {
			perfProfileActions[toName(profile)]->setChecked(true);
		},
		Qt::QueuedConnection);
}

void TrayIcon::setScheduler(std::optional<std::string> scheduler) {
	QMetaObject::invokeMethod(
		this,
		[=, this]() {
			schedulerActions[scheduler.value_or("")]->setChecked(true);
		},
		Qt::QueuedConnection);
}

void TrayIcon::setBatteryThreshold(BatteryThreshold threshold) {
	QMetaObject::invokeMethod(
		this,
		[=, this]() {
			thresholdActions[toString(threshold)]->setChecked(true);
		},
		Qt::QueuedConnection);
}

void TrayIcon::setBootSound(bool value) {
	QMetaObject::invokeMethod(
		this,
		[=, this]() {
			bootSoundActions[value]->setChecked(true);
		},
		Qt::QueuedConnection);
}

void TrayIcon::setProfileMenuEnabled() {
	QMetaObject::invokeMethod(
		this,
		[=, this]() {
			profileMenu->setEnabled(!onBattery && runningGames == 0);
		},
		Qt::QueuedConnection);
}

void TrayIcon::setCoolingMenuEnabled(bool enabled) {
	QMetaObject::invokeMethod(
		this,
		[=, this]() {
			coolingMenu->setEnabled(enabled);
		},
		Qt::QueuedConnection);
}

void TrayIcon::setSchedulerMenuEnabled() {
	QMetaObject::invokeMethod(
		this,
		[=, this]() {
			schedulerMenu->setEnabled(runningGames == 0);
		},
		Qt::QueuedConnection);
}

// ==============================
// Implementación de slots
// ==============================

void TrayIcon::onBatteryLimitChanged(BatteryThreshold value) {
	hardwareService.setChargeThreshold(value);
}

void TrayIcon::onEffectChanged(std::string effect) {
	openRgbService.setEffect(effect);
}

void TrayIcon::onBrightnessChanged(RgbBrightness brightness) {
	openRgbService.setBrightness(brightness);
}

void TrayIcon::onPerformanceProfileChanged(PerformanceProfile value) {
	if (steamService.getRunningGames().empty()) {
		performanceService.setPerformanceProfile(value);
	}
}

void TrayIcon::onSchedulerChanged(std::optional<std::string> scheduler) {
	if (steamService.getRunningGames().empty()) {
		performanceService.setScheduler(scheduler);
	}
}

// ==============================
// Constructor
// ==============================
TrayIcon::TrayIcon()
	: QObject(&MainWindow::getInstance()),
	  tray_icon_(new QSystemTrayIcon(this)),
	  tray_menu_(new QMenu(&MainWindow::getInstance())),
	  mainWindow(MainWindow::getInstance()) {
	tray_icon_->setIcon(QIcon::fromTheme(Constants::ASSET_ICON_FILE.c_str()));
	tray_icon_->setToolTip(Constants::APP_NAME.c_str());

	QMenu* menu = tray_menu_;

	onBattery	 = uPowerClient.isOnBattery();
	runningGames = steamService.getRunningGames().size();

	// -------------------------
	// Hardware menu
	// -------------------------

	if (hardwareService.getBatteryLimitAvailable() || hardwareService.getBootSoundAvailable()) {
		QAction* hardwareTitle = new QAction(translator.translate("hardware").c_str(), menu);
		hardwareTitle->setEnabled(false);
		menu->addAction(hardwareTitle);
		// -------------------------
		// BatteryThreshold submenu
		// -------------------------
		if (hardwareService.getBatteryLimitAvailable()) {
			QMenu* chargeLimitMenu	  = new QMenu(("    " + translator.translate("charge.threshold")).c_str(), menu);
			QActionGroup* chargeGroup = new QActionGroup(menu);
			for (BatteryThreshold bct : values<BatteryThreshold, true>()) {
				QAction* act = new QAction((std::to_string(toInt(bct)) + "%").c_str(), chargeGroup);
				act->setCheckable(true);
				act->setChecked(bct == hardwareService.getChargeThreshold());
				QObject::connect(act, &QAction::triggered, [this, bct]() {
					onBatteryLimitChanged(bct);
				});
				chargeLimitMenu->addAction(act);
				thresholdActions[toString(bct)] = act;
			}
			menu->insertMenu(nullptr, chargeLimitMenu);

			eventBus.onChargeThreshold([this](BatteryThreshold threshold) {
				setBatteryThreshold(threshold);
			});
		}
		// -------------------------
		// BatteryThreshold submenu
		// -------------------------
		// -------------------------
		// Boot sound submenu
		// -------------------------
		if (hardwareService.getBootSoundAvailable()) {
			QMenu* bootSoundMenu		 = new QMenu(("    " + translator.translate("boot.sound")).c_str(), menu);
			QActionGroup* bootSoundGroup = new QActionGroup(menu);

			const std::array<bool, 2> vals = {true, false};
			for (auto v : vals) {
				QAction* bsAct = new QAction(translator.translate(v ? "enabled" : "disabled").c_str(), bootSoundGroup);
				bsAct->setCheckable(true);
				bsAct->setChecked(v == hardwareService.getBootSound());
				QObject::connect(bsAct, &QAction::triggered, [this, v]() {
					onBootSoundChanged(v);
				});
				bootSoundMenu->addAction(bsAct);
				bootSoundActions[v] = bsAct;
			}

			menu->insertMenu(nullptr, bootSoundMenu);

			eventBus.onBootSound([this](bool value) {
				setBootSound(value);
			});
		}
		// -------------------------
		// Boot sound submenu
		// -------------------------
	}
	// -------------------------
	// Hardware menu
	// -------------------------

	menu->addSeparator();

	// -------------------------
	// Aura menu
	// -------------------------
	QAction* auraTitle = new QAction("Aura", menu);
	auraTitle->setEnabled(false);
	menu->addAction(auraTitle);
	// -------------------------
	// Effect submenu
	// -------------------------
	QMenu* effectMenu		  = new QMenu(("    " + translator.translate("effect")).c_str(), menu);
	QActionGroup* effectGroup = new QActionGroup(menu);
	auto effects			  = openRgbService.getAvailableEffects();
	for (std::string item : effects) {
		QAction* act = new QAction(item.c_str(), effectGroup);
		act->setCheckable(true);
		act->setChecked(item == openRgbService.getCurrentEffect());
		QObject::connect(act, &QAction::triggered, [this, item]() {
			onEffectChanged(item);
		});
		effectMenu->addAction(act);
		effectActions[item] = act;
	}
	menu->insertMenu(nullptr, effectMenu);

	eventBus.onRgbEffect([this](std::string effect) {
		setAuraEffect(effect);
	});
	// -------------------------
	// Effect submenu
	// -------------------------
	// -------------------------
	// Brightness submenu
	// -------------------------
	QMenu* brightnessMenu		  = new QMenu(("    " + translator.translate("brightness")).c_str(), menu);
	QActionGroup* brightnessGroup = new QActionGroup(menu);
	auto levels					  = values<RgbBrightness, true>();
	for (RgbBrightness item : levels) {
		QAction* act = new QAction(translator.translate("label.brightness." + toName(item)).c_str(), brightnessGroup);
		act->setCheckable(true);
		act->setChecked(item == openRgbService.getCurrentBrightness());
		QObject::connect(act, &QAction::triggered, [this, item]() {
			onBrightnessChanged(item);
		});
		brightnessMenu->addAction(act);
		brightnessActions[toName(item)] = act;
	}
	menu->insertMenu(nullptr, brightnessMenu);

	eventBus.onRgbBrightness([this](RgbBrightness brightness) {
		setAuraBrightness(brightness);
	});
	// -------------------------
	// Brightness submenu
	// -------------------------
	// -------------------------
	// Color submenu
	// -------------------------
	colorMenu = new QMenu(("    " + translator.translate("color")).c_str(), menu);
	colorMenu->setEnabled(openRgbService.getColor().has_value());

	currentColorAction = new QAction(StringUtils::toUpperCase(openRgbService.getColor().value_or("")).c_str(), menu);
	currentColorAction->setEnabled(false);
	colorMenu->addAction(currentColorAction);

	pickColorAction = new QAction((translator.translate("select")).c_str(), menu);
	QObject::connect(pickColorAction, &QAction::triggered, [this]() {
		QColor initial = QColor(openRgbService.getColor().value().c_str());
		QColor chosen  = QColorDialog::getColor(initial, nullptr, translator.translate("pick.color").c_str());

		if (chosen.isValid()) {
			openRgbService.setColor(chosen.name(QColor::HexRgb).toStdString());
		}
	});
	colorMenu->addAction(pickColorAction);

	menu->insertMenu(nullptr, colorMenu);

	eventBus.onRgbColor([this](std::optional<std::string> color) {
		setAuraColor(color);
	});
	// -------------------------
	// Color submenu
	// -------------------------
	// -------------------------
	// Aura menu
	// -------------------------

	menu->addSeparator();

	// -------------------------
	// Performance menu
	// -------------------------
	QAction* performanceTitle = new QAction(translator.translate("performance").c_str(), menu);
	performanceTitle->setEnabled(false);
	menu->addAction(performanceTitle);
	// -------------------------
	// Profile submenu
	// -------------------------
	profileMenu				   = new QMenu(("    " + translator.translate("profile")).c_str(), menu);
	QActionGroup* profileGroup = new QActionGroup(menu);
	auto items				   = values<PerformanceProfile>();
	for (PerformanceProfile item : items) {
		QAction* act = new QAction(translator.translate("label.profile." + toName(item)).c_str(), profileGroup);
		act->setCheckable(true);
		act->setChecked(item == performanceService.getPerformanceProfile());
		QObject::connect(act, &QAction::triggered, [this, item]() {
			onPerformanceProfileChanged(item);
		});
		profileMenu->addAction(act);
		perfProfileActions[toName(item)] = act;
	}

	profileMenu->setEnabled(runningGames == 0 && !onBattery);
	menu->insertMenu(nullptr, profileMenu);
	// -------------------------
	// Profile submenu
	// -------------------------
	// -------------------------
	// Scheduler submenu
	// -------------------------
	if (!performanceService.getAvailableSchedulers().empty()) {
		schedulerMenu			 = new QMenu(("    " + translator.translate("scheduler")).c_str(), menu);
		QActionGroup* schedGroup = new QActionGroup(menu);

		QAction* act = new QAction(performanceService.getDefaultSchedulerName().c_str(), schedGroup);
		act->setCheckable(true);
		act->setChecked(std::nullopt == performanceService.getCurrentScheduler());
		QObject::connect(act, &QAction::triggered, [this]() {
			onSchedulerChanged(std::nullopt);
		});
		schedulerMenu->addAction(act);
		schedulerActions[""] = act;
		schedulerMenu->addSeparator();

		schedulerMenu->setEnabled(runningGames == 0);

		auto items2 = performanceService.getAvailableSchedulers();
		std::reverse(items2.begin(), items2.end());
		for (auto item : items2) {
			act = new QAction(StringUtils::capitalize(item).c_str(), schedGroup);
			act->setCheckable(true);
			act->setChecked(item == performanceService.getCurrentScheduler());
			QObject::connect(act, &QAction::triggered, [this, item]() {
				onSchedulerChanged(item);
			});
			schedulerMenu->addAction(act);
			schedulerActions[item] = act;
		}
		menu->insertMenu(nullptr, schedulerMenu);

		eventBus.onScheduler([this](std::optional<std::string> scheduler) {
			setScheduler(scheduler);
		});
	}
	// -------------------------
	// Scheduler submenu
	// -------------------------
	// -------------------------
	// Fan curves menu
	// -------------------------
	if (!performanceService.getFans().empty()) {
		coolingMenu = new QMenu(("    " + translator.translate("fan.curves")).c_str(), menu);
		menu->insertMenu(nullptr, coolingMenu);

		QAction* act = new QAction((translator.translate("edit.curve") + "...").c_str());
		QObject::connect(act, &QAction::triggered, [this]() {
			openFanEditor();
		});
		coolingMenu->addAction(act);
		coolingMenu->setEnabled(performanceService.getPerformanceProfile() != PerformanceProfile::SMART);
	}
	// -------------------------
	// Fan curves menu
	// -------------------------
	// -------------------------
	// Game submenu
	// -------------------------
	QMenu* gamesMenu = new QMenu(("    " + translator.translate("games")).c_str(), menu);
	menu->insertMenu(nullptr, gamesMenu);

	QAction* act = new QAction((translator.translate("label.game.configure") + "...").c_str());
	QObject::connect(act, &QAction::triggered, [this]() {
		openGameList();
	});
	gamesMenu->addAction(act);

	// -------------------------
	// Game submenu
	// -------------------------
	// -------------------------
	// Performance menu
	// -------------------------

	menu->addSeparator();

	// -------------------------
	// Logs submenu
	// -------------------------
	QMenu* logsMenu = new QMenu(translator.translate("logs").c_str(), menu);
	menu->insertMenu(nullptr, logsMenu);

	QAction* openLogsAct = new QAction(translator.translate("open.log").c_str());
	QObject::connect(openLogsAct, &QAction::triggered, [this]() {
		shell.run_command("xdg-open " + Constants::LOG_DIR + "/" + Constants::LOG_FILE_NAME + ".log", false);
	});
	logsMenu->addAction(openLogsAct);

	QAction* openRunnerLogsAct = new QAction(translator.translate("open.log.runner").c_str());
	QObject::connect(openRunnerLogsAct, &QAction::triggered, [this]() {
		shell.run_command("xdg-open " + Constants::LOG_DIR + "/" + Constants::LOG_RUNNER_FILE_NAME + ".log", false);
	});
	logsMenu->addAction(openRunnerLogsAct);

	QAction* openOpenRGBLogsAct = new QAction(translator.translate("open.log.openrgb").c_str());
	QObject::connect(openOpenRGBLogsAct, &QAction::triggered, [this]() {
		shell.run_command("xdg-open " + Constants::LOG_DIR + "/" + Constants::LOG_ORGB_FILE_NAME + ".log", false);
	});
	logsMenu->addAction(openOpenRGBLogsAct);
	// -------------------------
	// Logs submenu
	// -------------------------

	menu->addSeparator();

	// -------------------------
	// Main window
	// -------------------------

	QAction* openAction = new QAction(translator.translate("open.ui").c_str(), menu);
	QObject::connect(openAction, &QAction::triggered, [this]() {
		openMainWindow();
	});
	menu->addAction(openAction);

	menu->addSeparator();

#ifdef DEV_MODE
	// -------------------------
	// Settings submenu
	// -------------------------
	QMenu* settingsMenu = new QMenu("Settings", menu);
	menu->insertMenu(nullptr, settingsMenu);

	QAction* openSettingsAct = new QAction("Open settings");
	QObject::connect(openSettingsAct, &QAction::triggered, [this]() {
		shell.run_command("xdg-open " + Constants::CONFIG_FILE);
	});
	settingsMenu->addAction(openSettingsAct);

	QAction* reloadSettingsAct = new QAction("Reload settings");
	QObject::connect(reloadSettingsAct, &QAction::triggered, [this]() {
		configuration.loadConfig();
	});
	settingsMenu->addAction(reloadSettingsAct);
	// -------------------------
	// Settings submenu
	// -------------------------
#endif

	// -------------------------
	// Main window
	// -------------------------

	menu->addSeparator();

	// -------------------------
	// Exit
	// -------------------------
	QAction* quitAction = new QAction(translator.translate("close").c_str(), menu);
	QObject::connect(quitAction, &QAction::triggered, [this]() {
		applicationService.shutdown();
	});
	menu->addAction(quitAction);
	// -------------------------
	// Exit
	// -------------------------

	tray_icon_->setContextMenu(menu);

	QObject::connect(tray_icon_, &QSystemTrayIcon::activated, this, [this](QSystemTrayIcon::ActivationReason reason) {
		if (reason == QSystemTrayIcon::Trigger) {
			openMainWindow();
		}
	});

	eventBus.onPerformanceProfile([this](PerformanceProfile profile) {
		setPerformanceProfile(profile);
		setCoolingMenuEnabled(profile != PerformanceProfile::SMART);
	});

	eventBus.onGameEvent([this](size_t runningGames) {
		this->runningGames = runningGames;
		setProfileMenuEnabled();
		setSchedulerMenuEnabled();
	});

	eventBus.onBattery([this](bool onBattery) {
		this->onBattery = onBattery;
		setProfileMenuEnabled();
	});
}

void TrayIcon::show() {
	tray_icon_->show();
}

void TrayIcon::onBootSoundChanged(bool enable) {
	hardwareService.setBootSound(enable);
}